File: iOS/Paging-CRUD/Paging-CRUD/Data/PostDao.swift
//
//  PostDao.swift
//  Paging-CRUD
//

import Foundation

// Post Caching (Single Source of Truth)
final class PostDao {
    static let shared = PostDao()

    private var cachedMap = [Int: [ListItem.Post]]()

    private var invalidationListeners = [UUID: () -> Void]()

    private let lock = NSLock()

    private init() {}

    @discardableResult
    func addInvalidationListener(_ listener: @escaping () -> Void) -> UUID {
        lock.lock()
        defer { lock.unlock() }
        let token = UUID()

        invalidationListeners[token] = listener
        return token
    }

    func removeInvalidationListener(_ token: UUID) {
        lock.lock()
        defer { lock.unlock() }
        invalidationListeners[token] = nil
    }

    private func notifyInvalidated() {
        lock.lock()
        let listeners = Array(invalidationListeners.values)
        lock.unlock()
        listeners.forEach { $0() }
    }

    func insertAll(_ key: Int, _ list: [ListItem.Post]) {
        lock.lock()
        let cachedIds = Set(cachedMap[key, default: []].map { $0.id })

        cachedMap[key, default: []].append(contentsOf: list.filter { !cachedIds.contains($0.id) })
        lock.unlock()
        notifyInvalidated()
    }

    func replaceAll(_ key: Int, _ list: [ListItem.Post]) {
        lock.lock()
        cachedMap[key] = list
        lock.unlock()
        notifyInvalidated()
    }

    func getPostList(_ key: Int, _ start: Int, _ end: Int) -> [ListItem.Post] {
        lock.lock()
        defer { lock.unlock() }
        guard let list = cachedMap[key], start < list.count else { return [] }
        return Array(list[start..<min(end, list.count)])
    }

    func deletePost(_ postId: Int) {
        lock.lock()
        guard let key = cachedMap.first(where: { $0.value.contains { $0.id == postId } })?.key else {
            lock.unlock()
            return
        }

        cachedMap[key]?.removeAll { $0.id == postId }
        lock.unlock()
        notifyInvalidated()
    }

    func deleteAll(_ key: Int) {
        lock.lock()
        cachedMap[key]?.removeAll()
        lock.unlock()
        notifyInvalidated()
    }

    func isCacheEmpty(_ key: Int) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return cachedMap[key, default: []].isEmpty
    }

    func getCount(_ key: Int) -> Int {
        lock.lock()
        defer { lock.unlock() }
        return cachedMap[key]?.count ?? 0
    }
}

File: Paging-CRUD-main/shared/src/commonMain/kotlin/com/hhp227/paging_crud/data/PostDao.kt
package com.hhp227.paging_crud.data

import com.hhp227.paging_crud.model.ListItem
import kotlinx.atomicfu.locks.SynchronizedObject
import kotlinx.atomicfu.locks.synchronized

// Post Caching (Single Source of Truth)
// JVM 전용 ConcurrentHashMap 대신 atomicfu 락으로 KMP 공통 스레드 안전성을 확보한다
object PostDao {
    private val lock = SynchronizedObject()

    private val cachedMap = mutableMapOf<Int, MutableList<ListItem.Post>>()

    private val invalidationListeners = mutableListOf<() -> Unit>()

    fun addInvalidationListener(listener: () -> Unit) {
        synchronized(lock) { invalidationListeners += listener }
    }

    fun removeInvalidationListener(listener: () -> Unit) {
        synchronized(lock) { invalidationListeners -= listener }
    }

    private fun notifyInvalidated() {
        val listeners = synchronized(lock) { invalidationListeners.toList() }

        listeners.forEach { it() }
    }

    fun insertAll(key: Int, list: List<ListItem.Post>) {
        synchronized(lock) {
            val posts = cachedMap.getOrPut(key) { mutableListOf() }
            val cachedIds = posts.map { it.id }.toSet()

            posts += list.filter { it.id !in cachedIds }
        }
        notifyInvalidated()
    }

    fun replaceAll(key: Int, list: List<ListItem.Post>) {
        synchronized(lock) { cachedMap[key] = list.toMutableList() }
        notifyInvalidated()
    }

    fun getPostList(key: Int, start: Int, end: Int): List<ListItem.Post> {
        return synchronized(lock) {
            cachedMap[key]?.let { list ->
                if (start >= list.size) emptyList()
                else list.slice(start until if (end < list.size) end else list.size)
            } ?: emptyList()
        }
    }

    fun deletePost(postId: Int) {
        val removed = synchronized(lock) {
            val key = cachedMap.entries.find { entry -> entry.value.any { it.id == postId } }?.key
                ?: return

            cachedMap[key]?.removeAll { it.id == postId } == true
        }

        if (removed) {
            notifyInvalidated()
        }
    }

    fun deleteAll(key: Int) {
        synchronized(lock) { cachedMap[key]?.clear() }
        notifyInvalidated()
    }

    fun isCacheEmpty(key: Int) = synchronized(lock) { cachedMap[key]?.isEmpty() ?: true }

    fun getCount(key: Int) = synchronized(lock) { cachedMap[key]?.size ?: 0 }
}
